contracts/amm/WTokenVault.sol

```diff
@@ -0,0 +1,261 @@
+                    addressesProvider.getAmmDataProvider()
+                ).getPriceForSeries(seriesId, vars.volatility);
+
+                uint256 valuePerToken;
+                if (series.isPutOption) {
+                    valuePerToken = seriesController.getCollateralPerUnderlying(
+                            seriesId,
+                            (series.strikePrice * 1e18) /
+                                vars.underlyingPrice -
+                                bPrice,
+                            vars.underlyingPrice
+                        );
+                } else {
+                    valuePerToken = 1e18 - bPrice;
+                }
+
+                uint256 poolWTokenBalance = lockedWTokens[address(amm)][
+                    seriesId
+                ];
+                if (poolWTokenBalance > 0) {
+                    vars.poolValue[expirationId - vars.expirationIdMin] +=
+                        (poolWTokenBalance * valuePerToken) /
+                        1e18;
+                }
+
+                if (wTokenAmount > 0) {
+                    // Increase locked wToken amount
+                    lockedWTokens[address(amm)][seriesId] += wTokenAmount;
+                    vars.lockedValue[expirationId - vars.expirationIdMin] +=
+                        (wTokenAmount * valuePerToken) /
+                        1e18;
+                }
+            }
+        }
+
+        for (
+            uint64 i = 0;
+            i <= vars.expirationIdMax - vars.expirationIdMin;
+            i++
+        ) {
+            if (vars.lockedValue[i] > 0) {
+                uint256 expirationId = i + vars.expirationIdMin;
+                // Add locked collateral to the expiration ID
+                vars.poolValue[i] += lockedCollateral[address(amm)][
+                    expirationId
+                ];
+
+                // Update LP shares balance and supply
+                uint256 existingSupply = lpSharesSupply[address(amm)][
+                    expirationId
+                ];
+                uint256 newSupply;
+                if (existingSupply == 0) {
+                    newSupply = vars.lockedValue[i];
+                } else {
+                    newSupply =
+                        ((vars.poolValue[i] + vars.lockedValue[i]) *
+                            existingSupply) /
+                        vars.poolValue[i];
+                }
+
+                lpShares[address(amm)][expirationId][redeemer] +=
+                    newSupply -
+                    existingSupply;
+                lpSharesSupply[address(amm)][expirationId] = newSupply;
+
+                emit WTokensLocked(
+                    address(amm),
+                    redeemer,
+                    expirationId,
+                    newSupply - existingSupply
+                );
+            }
+        }
+    }
+
+    /// Redeem locked collateral post-expiration
+    function redeemCollateral(uint256 expirationId, address redeemer)
```

Comment:
Do we want non reentrant on this? Is it possible to have a reentrancy attack on this contract even though the caller is protected by reentrancy?

Reply:
no need, because there's no token movement here, only accounting. The AMM has nonReentrant on the method that calls this